Auto-generate infix declarations

int-index opened the pull request:
closes #157

goldfirere commented on src/Data/Singletons/Single/Fixity.hs:
```diff
@@ -0,0 +1,27 @@
+  | isUpcase name =
+    -- is it a tycon name or a datacon name??
+    -- it *must* be a datacon name, because symbolic tycons
+    -- can't be promoted. This is terrible.
```
This comment is out-of-date, because `TypeInType` allows you to promote symbolic tycons and datacons. Fixing this would take Some Work, unfortunately, because you'd have to poke around to see what's in scope. In any case, it's orthogonal to your patch.

goldfirere commented on src/Data/Singletons/Single/Fixity.hs:
```diff
@@ -0,0 +1,27 @@
+
+singFixityDeclaration :: DsMonad q => Name -> q [DDec]
+singFixityDeclaration name = do
+  mFixity <- qReifyFixity name
```
What will this do if `name` is out of scope? It doesn't fail, does it?

goldfirere commented on src/Data/Singletons/Single/Data.hs:
```diff
@@ -30,7 +31,8 @@ singDataD (DataDecl _nd name tvbs ctors derivings) = do
   k <- promoteType (foldType (DConT name) (map DVarT tvbNames))
   ctors' <- mapM (singCtor a) ctors
-
+  ctorFixities <- singFixityDeclarations
```
That this works is a delicate balance that should probably be spelled out.

If we are singling a previously defined datatype (whose definition is retrieved by reification), then the `qReifyFixity` calls underneath `singFixityDeclarations` work. If we are singling a datatype being defined right now (e.g., in a call to `singletons`) then this works because we will also encounter the fixity declaration and single it separately. But wait: does the `DsMonad` machinery succeed for `qReifyFixity` in this case? Would we then get two singled fixity declarations?

goldfirere commented:
Thanks much for putting this together. It looks plausible. But we can't be sure without a test! Make sure to test with `genSingletons` and `singletons`, both symbolic names and alphanumeric names. Again, thanks for playing!

int-index commented:
Thanks for a review! I'll continue working on this tomorrow.

int-index commented:
> tomorrow

ha ha :confused: 

Anyway, I made three changes as discussed:
1. fixed the problem with out-of-scope names that cannot be reified
2. added a comment that explains the approach used
3. added a test with `singletons`/`genSingletons` and both symbolic and alphanumeric names

goldfirere commented:
Yay. Thanks!
